internal/sci/gcp/manager.go

```diff
@@ -0,0 +1,188 @@
+// Package gcp provides a GCP implementation of the Substratus Cloud Interface (SCI).
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	credentials "cloud.google.com/go/iam/credentials/apiv1"
+	credentialspb "cloud.google.com/go/iam/credentials/apiv1/credentialspb"
+	"cloud.google.com/go/storage"
+	"github.com/sethvargo/go-envconfig"
+	"github.com/substratusai/substratus/internal/sci"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/iam/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// server implements the sci.ControllerServer interface.
+type Server struct {
+	sci.UnimplementedControllerServer
+	Clients
+	SaEmail   string
+	ProjectID string `env:"PROJECT_ID"`
+}
+
+type Clients struct {
+	IAMCredentialsClient *credentials.IamCredentialsClient
+	IAM                  *iam.Service
+	Metadata             *metadata.Client
+	Storage              *storage.Client
+	Http                 *http.Client
+}
+
+func NewServer() (*Server, error) {
+	s := &Server{}
+	ctx := context.Background()
+	if err := envconfig.Process(ctx, s); err != nil {
+		return nil, fmt.Errorf("environment: %w", err)
+	}
+	return s, nil
+}
+
+// CreateSignedURL generates a signed URL for a specified GCS bucket and object path.
+func (s *Server) CreateSignedURL(ctx context.Context, req *sci.CreateSignedURLRequest) (*sci.CreateSignedURLResponse, error) {
+	log := log.FromContext(ctx)
+	log.Info("creating signed URL", "bucket", req.BucketName, "object", req.ObjectName)
+
+	bucketName, objectName, checksum := req.GetBucketName(),
+		req.GetObjectName(),
+		req.GetMd5Checksum()
+	bucket := s.Clients.Storage.Bucket(bucketName)
+	obj := bucket.Object(objectName)
+	if _, err := obj.Attrs(ctx); err != nil && err != storage.ErrObjectNotExist {
```

Comment:
What is this used for?

Reply:
I think Brandon did this to check for errors by seeing if it worked to get attrs? @brandonjbjelland might be able to explain. This is coming over from original gcpmanager code.

Reply:
My thought process was that it was a good idea to check what's at the current path and what we should expect is that no object exists. If we encounter some other error (permissions or whatever), I think it's a good indication that we wouldn't be able to create the signed URL or have it be used without error by the uploader.

At some point I think we should change how we do uploading to include the MD5checksum in the path, and verify that an object doesn't exist here prior to creating an upload URL (if it does, it means we'd be doing a re-upload and we can likely short circuit the upload process). As it currently stands, if I do two notebook uploads in rapid succession (with some change in between) or if two people are working on the same notebook and performing uploads, I think the state is indeterminate.

Reply:
I'm going to ignore this since changing the logic of signedURL wasn't in scope of this PR. Feel free to open a new issue and our PM can consider prioritization of it.

Reply:
@nstogner ☝️  thoughts?